service/matching/version_rule_helpers.go

```diff
@@ -248,3 +341,35 @@ func CleanupRuleTombstones(versioningData *persistencepb.VersioningData, retenti
 	})
 	return modifiedData
 }
+
+// CommitBuildID makes the following changes:
+//  1. Adds an unconditional assignment rule for the target Build ID at the
+//     end of the list. An unconditional assignment rule:
+//     - Has no hint filter
+//     - Has no ramp
+//  2. Removes all previously added assignment rules to the given target
+//     Build ID (if any).
+//  3. Removes any *unconditional* assignment rule for other Build IDs.
+func CommitBuildID(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId) (*persistencepb.VersioningData, error) {
+	data = common.CloneProto(data)
+	target := req.GetTargetBuildId()
+	assignmentRules := data.GetAssignmentRules()
+	for _, ar := range getActiveAssignmentRules(assignmentRules) {
+		if ar.GetRule().GetTargetBuildId() == target {
+			ar.DeleteTimestamp = timestamp
+		}
+		if isUnfiltered(ar.GetRule()) {
+			ar.DeleteTimestamp = timestamp
+		}
+	}
+	data.AssignmentRules = append(assignmentRules, &persistencepb.AssignmentRule{
+		Rule:            &taskqueue.BuildIdAssignmentRule{TargetBuildId: target},
+		CreateTimestamp: timestamp,
+	})
+	if err := checkAssignmentConditions(data, 0, false, true); err != nil {
```

Comment:
hmm... this check does not seem right, or necessary to me.

Reply:
this check will make sure that the build id being committed isn't part of an existing version set, which I think is necessary, but actually it will be more efficient if we check that only for the build id that's being committed, and do the check at the beginning of the function. If `hadUnfiltered == true`, then `checkAssignmentConditions` also makes sure there's still an unfiltered rule there, but we shouldnt check that because we know it's true.

Reply:
the only relevant part of the check is to see whether adding the new rule causes the number of rules to exceed the max, which could happen if no rules are deleted. I changed the checkAssignmentConditions fn to be more efficient and only check the conditions that it needs to